util: add RunCommandParseJSON

Sjors opened the pull request:
Prerequisite for external signer support in #16546. Big picture overview in [this gist](https://gist.github.com/Sjors/29d06728c685e6182828c1ce9b74483d).

This adds a new dependency [boost process](https://github.com/boostorg/process/tree/boost-1.64.0). This is part of Boost since 1.64 which is part of `depends`. Because the minimum Boost version is 1.47, this functionality is skipped for older versions of Boost.

Use `./configure --with-boost-process` to opt in, which checks for the presence of Boost::Process.

We add `UniValue runCommandParseJSON(const std::string& strCommand)` to `system.{h,cpp}` which calls an arbitrary command and processes the JSON returned by it. This is currently only called by the test suite.

~For testing purposes this adds a new regtest-only RPC method `runcommand`, as well as `test/mocks/command.py` used by functional tests.~ (this is no longer the case)

TODO:
- [ ] review boost process in #15440

Sjors commented:
See earlier discussion in https://github.com/bitcoin/bitcoin/pull/14912#pullrequestreview-186165666 for different approaches like `_popen` or `boost process`.

cc @practicalswift @promag @ken2812221 @ryanofsky

jgarzik commented:
This C++11 python-inspired subprocess module may or may not be of relevant interest, as it avoids Boost in favor of straight C++11: http://templated-thoughts.blogspot.com/2016/03/sub-processing-with-modern-c.html

Sjors commented:
@jgarzik thanks! [1600 lines of code](https://github.com/arun11299/cpp-subprocess/blob/master/subprocess.hpp) seems a bit overkill, but on the other hand it looks very clean and presumably better than anything I would come up with.

Speaking of UniValue :-) I think I need to reorder the Makefile to prevent some Travis instances from failing with `undefined reference to `UniValue::read(char const*, unsigned int)`.

practicalswift commented:
The `-regtest` requirement is very important: could add a test case triggering the 
"runcommand for regression testing (`-regtest` mode) only" code path?

Sjors commented:
@practicalswift agreed, but I don't know if the functional test framework can launch a Testnet node without triggering a real IBD. I suppose I can configure it to not connect to any peers.

MarcoFalke commented on src/rpc/misc.cpp:
```diff
@@ -364,6 +364,35 @@ static UniValue setmocktime(const JSONRPCRequest& request)
+    if (request.fHelp || request.params.size() != 1)
+        throw std::runtime_error(
+            RPCHelpMan{"runcommand",
+                "\nRun command and parse JSON from stdout (-regtest only)\n",
```
Why is this RPC exposed? For testing purposes? It should be a unit test in that case.

Sjors replied:
It's restricted to regtest. Main reasons I didn't build a unit test:
1. no idea how
2. potentially creates even more cross-platform headaches
3. `runCommand` is also only tested via functional tests (via `blocknotify`, etc)

promag replied:
> potentially creates even more cross-platform headaches

What you mean? Because of the command itself?

promag replied:
@Sjors how about supporting `-exec "cmd"` in bitcoin_test binary (or in any other test binary), then that could be used in the functional test (instead of this RPC).

ryanofsky replied:
I think it's ok to have this RPC only enabled in regtest. If you wanted to add more protection against misuse, you could disable it without `gArgs.GetBoolArg("-enable-unsafe-runcommand-rpc", false)` or similar.

If you did want to drop the RPC and just write a unit test for `runCommandParseJSON`, I don't think it would be difficult to do by just having the unit test write a temporary json file and then calling `runCommandParseJSON` with `cat path_to_file` on unix and `type path_to_file` on windows.

Sjors commented:
AppVeyor seems unhappy about the `RPCHelpMan` syntax:

```
[C:\projects\bitcoin\build_msvc\libbitcoin_util\libbitcoin_util.vcxproj]
25c:\projects\bitcoin\src\leveldb\db\skiplist.h(360): warning C4312: 'reinterpret_cast': conversion from 'int' to 'void *' of greater size [C:\projects\bitcoin\build_msvc\libleveldb\libleveldb.vcxproj]
26c:\projects\bitcoin\src\rpc\misc.cpp(371): error C2440: 'initializing': cannot convert from 'initializer list' to 'RPCHelpMan' [C:\projects\bitcoin\build_msvc\libbitcoin_server\libbitcoin_server.vcxproj]
27c:\projects\bitcoin\src\rpc\misc.cpp(379): error C2512: 'std::runtime_error': no appropriate default constructor available [C:\projects\bitcoin\build_msvc\libbitcoin_server\libbitcoin_server.vcxproj]
28Command exited with code 1
29
```

practicalswift commented on test/functional/mocks/command.py:
```diff
@@ -0,0 +1,32 @@
+parser.set_defaults(func=command)
+
+if len(sys.argv) == 1:
+   args = parser.parse_args(['-h'])
```
Wrong indentation :-)

practicalswift commented on test/functional/mocks/command.py:
```diff
@@ -0,0 +1,32 @@
+import json
+import sys
+
+def command(args):
```
This is redefining `args` from the outer scope.

MarcoFalke commented on src/rpc/misc.cpp:
```diff
@@ -364,6 +364,35 @@ static UniValue setmocktime(const JSONRPCRequest& request)
+            RPCHelpMan{"runcommand",
+                "\nRun command and parse JSON from stdout (-regtest only)\n",
+                {
+                    {"command", RPCArg::Type::STR, /* opt */ false, /* default_val */ "", "Command must return JSON\n"},
```
```suggestion
                    {"command", RPCArg::Type::STR, RPCArg::Optional::NO, "Command must return JSON\n"},
```

Sjors committed:
[build] msvc: add boost::process

* AppVeyor boost-process vcpkg package.
* Tell Boost linter to ignore it
* Add HAVE_BOOST_PROCESS for MSVC build (bitcoin_config.h)

Sjors commented:
I switched to boost::process, which should give us Windows support. It's included with the depends Boost 1.64. Because the minimum Boost version is 1.47, this functionality is just not compiled for older versions of Boost.

Sjors commented:
@DrahtBot AppVeyor `vcpkg install` [fails](https://ci.appveyor.com/project/DrahtBot/bitcoin/builds/22349441) (I tried restarting), mostly likely as a result of me adding the `boost-process` package:
<img width="938" alt="schermafbeelding 2019-02-13 om 21 20 30" src="https://user-images.githubusercontent.com/10217/52741261-3e396880-2fd5-11e9-8d6f-19d0b034df6b.png">

I add AppVeyor to my own Github account and it finished `vcpkg install` [successfully](https://ci.appveyor.com/project/Sjors/bitcoin) (it's still building though).

@MarcoFalke maybe it needs a cache purge?

MarcoFalke commented:
Not sure how to purge the cache, maybe @ken2812221 knows

Sjors commented:
On the bright side, the Gitian build does work:
<img width="515" alt="schermafbeelding 2019-02-13 om 21 57 48" src="https://user-images.githubusercontent.com/10217/52743539-8b6c0900-2fda-11e9-94a8-278488d0b454.png">

Sjors commented:
Trying something based on the [docs](https://www.appveyor.com/docs/build-cache/#cleaning-up-cache)...

Sjors commented:
I can't build master either on my own instance ([logs](https://ci.appveyor.com/project/Sjors/bitcoin/builds/22351406)).

```
...
libbitcoin_server.lib(index_txindex.obj) : error LNK2001: unresolved external symbol "public: void __cdecl leveldb::WriteBatch::Clear(void)" (?Clear@WriteBatch@leveldb@@QEAAXXZ) [C:\projects\bitcoin\build_msvc\bitcoin-wallet\bitcoin-wallet.vcxproj]
53C:\projects\bitcoin\build_msvc\x64\Release\bitcoin-wallet.exe : fatal error LNK1120: 14 unresolved externals [C:\projects\bitcoin\build_msvc\bitcoin-wallet\bitcoin-wallet.vcxproj]
```

Are there any non-default settings I (and the docs) should know about?

jonasschnelli commented:
@Sjors:
I miss some context. Can you elaborate why this is necessary for external signer support?

Sjors commented:
@jonasschnelli #14912 allows Bitcoin Core to communicate with [HWI](https://github.com/bitcoin-core/HWI) which in turns communicates with the device. That way we don't have to implement device specific support, but instead communicate via a (to be) standardised protocol to a driver. That driver could be HWI, but it could be any other command line utility that understands the same commands and gives the same responses. The protocol is described in more detail [here](https://github.com/Sjors/bitcoin/blob/2018/11/rpc-signer/doc/external-signer.md#signer-api).

ken2812221 commented:
Looks like appveyor has not updated their `boost-process` package.
Maybe you could try to add
```yml
- cmd: git -C C:\tools\vcpkg pull -q
```
before
```yml
- cmd: vcpkg remove --outdated --recurse
```
But I haven't tested it.

Sjors commented:
@ken2812221 AppVeyor actually passed before my rebase. I think the problem is that it needs to bust cache before running, so that it adds `boost-process`.

luke-jr commented:
Most distros seem to split boost modules, so just because 1.64 is used doesn't guarantee boost::process is installed I think? We probably need a configure check...

Sjors commented:
@luke-jr good point, I actually noticed that with AppVeyor which aborted much too late in the build process before I added the `boost-process` package.

Sjors commented:
Rebased. I cherry-picked Luke's commit to explicitly detect boost-process with `#ifdef HAVE_BOOST_PROCESS`.

I started reviewing boost-process in #15440

MarcoFalke commented on src/Makefile.test.include:
```diff
@@ -518,10 +518,10 @@ test_fuzz_blocktransactionsrequest_deserialize_CPPFLAGS = $(AM_CPPFLAGS) $(BITCO
  $(LIBBITCOIN_SERVER) \
  $(LIBBITCOIN_COMMON) \
  $(LIBBITCOIN_UTIL) \
+ $(LIBUNIVALUE) \
```
Needs rebase and all those ugly hunks dropped in this file

Sjors replied:
Hooray, I was able to drop 829171fa3182db545d9a07834fa1c2006fa01b6e

Sjors commented:
Added support for stdin.

Sjors commented:
Not sure what I'm doing wrong here...

<img width="1247" alt="Schermafbeelding 2019-03-15 om 15 10 34" src="https://user-images.githubusercontent.com/10217/54437257-85318f80-4734-11e9-9666-8833e96c9e95.png">

jgarzik commented:
> Not sure what I'm doing wrong for the windows cross compile [build](https://travis-

The string `stdin` is a symbol from `stdio.h` of type `FILE *`.  The compiler thinks you are passing the well-known FILE* to runCommandParseJSON()

Recommend avoiding the `stdin` string altogether, since that's a symbol from the earliest days of ANSI C.  Call it `stdinData` or something.

practicalswift commented on test/functional/test_framework/test_framework.py:
```diff
@@ -571,3 +576,7 @@ def is_zmq_compiled(self):
 
         return config["components"].getboolean("ENABLE_ZMQ")
+
+    def is_runcommand_compiled(self, node):
```
This does not have to be a method, right? Could be a function instead?

Sjors commented:
Note that on macOS with Homebrew boost 1.69.0_2 won't find `boost::process`, due to similar errors as described [here](https://github.com/boostorg/process/issues/55). It works fine with depends though.

promag commented on src/util/system.cpp:
```diff
@@ -1140,6 +1145,38 @@ void runCommand(const std::string& strCommand)
 #endif
 
+#ifdef HAVE_BOOST_PROCESS
+UniValue runCommandParseJSON(const std::string& strCommand, const std::string& strStdIn)
```
Will this be used out of RPC? Otherwise could be moved to src/rpc/util.h?

Sjors replied:
Yes, this be used by the wallet (including from the GUI).

promag commented on test/functional/mocks/command.py:
```diff
@@ -0,0 +1,40 @@
+        raise RuntimeError("Missing arguments")
+
+
+parser = argparse.ArgumentParser(prog='./mock_command.py', description='Test runCommandParseJSON() via runcommand RPC')
```
nit, wrong `prog=`.

promag reviewed (commented):
Tested locally on macos and looks good. I'll keep #13339 rebase with this to have boost process support - will keep testing on my branch. Nice work @Sjors and @luke-jr.

promag commented on src/util/system.h:
```diff
@@ -82,6 +84,16 @@ fs::path GetSpecialFolderPath(int nFolder, bool fCreate = true);
 #endif
+#ifdef HAVE_BOOST_PROCESS
+/**
+ * Execute a command with returns JSON, and parse the result.
```
Typo, witch.

Also, could say what happens if the output can't be parsed.

practicalswift replied:
> Typo, witch.

Which, seems like like a typo witch hunt.

promag replied:
Ah! I wanted to type "with".

hebasto replied:
A typo seems remain :)
```suggestion
 * Execute a command which returns JSON, and parse the result.
```

promag commented:
While testing #13339 rebased with this one I've come to the conclusion that `boost::process::system` is way slower than standard `system`. Looks like there is some internal synchronization when multiple concurrent calls are made - can be observed with `bitcoin-cli -regtest generatetoaddress 10 ...` and `-blocknotify=...`.

practicalswift commented on src/rpc/misc.cpp:
```diff
@@ -366,6 +366,40 @@ static UniValue setmocktime(const JSONRPCRequest& request)
+            }.ToString()
+        );
+
+    if (!Params().MineBlocksOnDemand())
```
Can we check for `-regtest` explicitly instead of implicitly via `fPowNoRetargeting`? Explicit feels better than implicit when it comes to enabling remote execution :)

Sjors replied:
Done

Sjors commented:
I added `HAVE_BOOST_PROCESS` to `test/config.ini` so that the test suite can check for it directly.  Previously the test suite would call `runcommand` to find out. I also added `--with-boost-process` to allow opting out, though I'm not sure how this interacts with the same flag in `ax_boost_process.m4`.

practicalswift commented:
@Sjors Since `runcommand` support increases the attack surface perhaps it should be opt-in (`--enable-external-signer`) instead of opt-out (`--disable-external-signer`)?

Sjors commented:
@practicalswift if this were a standalone PR I would agree, but this PR only makes sense if we decide to go ahead with external signer support. Disabling that by default seems overkill.

ryanofsky commented on src/util/system.cpp:
```diff
@@ -1148,6 +1153,38 @@ void runCommand(const std::string& strCommand)
 #endif
 
+#ifdef USE_EXTERNAL_SIGNER
+UniValue runCommandParseJSON(const std::string& strCommand, const std::string& strStdIn)
```
In commit "[util] add runCommandParseJSON" (82079a04e6505424bbcb4b030a81d368933d24a7)

Naming convention for argument and variables in this function could use snake_case instead of camelCase and drop strHungarianNotationPrefixes

ryanofsky commented on configure.ac:
```diff
@@ -125,6 +125,11 @@ AC_ARG_ENABLE([upnp-default],
   [use_upnp_default=no])
 
+AC_ARG_ENABLE([external-signer],
+    [AS_HELP_STRING([--disable-external-signer],[do not compile external signer support (default is to compile if Boost::Process is found)])],
```
In commit "configure: add --enable-external-signer" (ec6af5d1c993d584861be4ce6a0f5188f1af73af)

Before we actually check in code making this flag useful, the `./configure --help` text should say `(incomplete, experimental)` or similar to avoid suggesting the build will support a feature that isn't actually usable.

ryanofsky commented on configure.ac:
```diff
@@ -1640,33 +1667,34 @@ esac
 fi
-echo "  with zmq      = $use_zmq"
-echo "  with test     = $use_tests"
+echo "  external signer = $use_external_signer"
```
In commit "configure: add --enable-external-signer" (ec6af5d1c993d584861be4ce6a0f5188f1af73af)

This is pretty misleading to print before the feature actually implemented. Would suggest commenting this out by default to avoid confusion.

Sjors replied:
I commented this line out, but leaving `--enable-external-signer` in place (and keeping the indentation change).

ryanofsky reviewed (approved):
Code review ACK 82079a04e6505424bbcb4b030a81d368933d24a7

Looks great! Really nicely implemented updating everything from the msvc build to the test framework to the doxygen configuration. I left review suggestions that I think would be nice to address but are not critical.

ryanofsky commented:
> TODO:
> 
> * [ ]  review boost process in #15440

Just saw this TODO. Is it still valid? If so, could you clarify who is supposed to do what? Is this blocking? Or do you just want to encourage people to look at the boost process implementation?(Could include a link to https://github.com/boostorg/process in that case)

If Boost.Process does need more scrutiny, that might be a reason to disable its use by default. It might also be a reason to replace or augment the `--enable-external-signer` option with a `--with-boost-process` flag for clarity. A `--with-boost-process` flag might also be nice because Boost.Process could be used for other things not related to external signing, like wallet notifications (#13339).

My previous review ack / approval still stands, but I am wondering about this TODO in the PR description.

ryanofsky commented:
> If Boost.Process does need more scrutiny, that might be a reason to disable its use by default.

Note that to disable boost process by default but still get test coverage on travis you would need to add a depends option and enable the option it in one or more travis configurations, passing it to `configure` through [`depends/config.site.in`](https://github.com/bitcoin/bitcoin/blob/feb1a8c03aff0d37730d80eee7a05cc7e4729850/depends/config.site.in#L20-L34). #16367 has an example of this with the `MULTIPROCESS` depends flag and `--enable-multiprocess` autoconf option.

Sjors commented on build-aux/m4/ax_boost_process.m4:
```diff
@@ -0,0 +1,114 @@
+#
+# LICENSE
+#
+#   Copyright (c) 2012 Xiyue Deng <manphiz@gmail.com>
```
@luke-jr?

Sjors replied:
I swapped it out in favor of a more recent boost::system based example, listed the changes, and made an upstream PR: https://github.com/autoconf-archive/autoconf-archive/pull/202

Sjors commented on configure.ac:
```diff
@@ -1451,6 +1457,26 @@ else
   fi
 fi
 
+dnl enable external signer support
```
I should also rename this to `--disable-external-signer` given the default is enable when a recent enough Boost version is found. I'll update the version info as well, although I believe it will throw if you try `--enable-external-signer` on a old version of Boost.

Sjors replied:
Added, and clarified the commit to indicate external signer is an upcoming feature.

Sjors commented on configure.ac:
```diff
@@ -1451,6 +1457,26 @@ else
+  if test x$use_external_signer != xno; then
+    AC_MSG_RESULT(yes)
+    use_external_signer=yes
+    AC_DEFINE_UNQUOTED([USE_EXTERNAL_SIGNER],[$use_external_signer],[External signer support not compiled if undefined, otherwise value (0 or 1) determines default state])
```
Any tips for efficiently renaming a constant accross multiple commits?

Sjors replied:
To rename in each commit:
```sh
git rebase -i HEAD~5 --exec 'git diff --name-only --diff-filter=AMC HEAD~ | xargs gsed -i "s/USE_EXTERNAL_SIGNER/ENABLE_EXTERNAL_SIGNER/g"; git commit -a --amend --no-edit'
```

To rename messages:
```sh
git filter-branch -f --msg-filter 'gsed -e "s/USE_EXTERNAL_SIGNER/ENABLE_EXTERNAL_SIGNER/g"'  HEAD~5..HEAD
```

I wasn't able to get `filter-branch` to work with `--tree-filter` like here: https://stackoverflow.com/questions/39608191/rename-variable-in-all-commits-on-a-given-branch (e.g. this leads to followup commits undoing earlier ones: `git filter-branch -f --tree-filter 'git diff --name-only --diff-filter=AMC $GIT_COMMIT~ | xargs gsed -i "s/USE_EXTERNAL_SIGNER/ENABLE_EXTERNAL_SIGNER/g"' HEAD~4^..HEAD HEAD`)

Sjors committed:
configure: add ax_boost_process

Co-authored-by: Luke Dashjr <luke-jr+git@utopios.org>

Sjors commented:
Rebased, implemented some of @ryanofsky's suggestions. I might look into ditching the `runcommand` test RPC in favor of a unit test later.

laanwj commented:
I haven't dived into details of why it's needed for external signers, but from a security perspective I really dislike an RPC call that can invoke arbitrary commands.

Sjors commented:
> I haven't dived into details of why it's needed for external signers, but from a security perspective I really dislike an RPC call that can invoke arbitrary commands.
> 
> Edit: could we restrict this to a set of commands specified in `bitcoin.conf`?

External signer support has precisely that restriction (the only command that's called is specified by `-signer`).

Sjors commented:
I dropped the `runcommand` regtest RPC call and added `test/system_tests.cpp` instead. This test file creates the `command.py` test script programatically, because `test_bitcoin` is a standalone executable.

Sjors committed:
[doc] include Doxygen comments for HAVE_BOOST_PROCESS

Sjors commented:
I simplified the tests by just using `echo` instead of calling a Python script.

I moved the external signer specific commits to #16546. `runCommandParseJSON` now checks for `HAVE_BOOST_PROCESS`.

Sjors commented:
Rebased after native ARM support for Travis was added in #17233.  I re-enabled Boost::Process for the ARM Travis machine (replacing 23b653aad with 31c2d4b84fd9292387adbda9a4dbfdd0ca83645).

Sjors commented:
@sipsorcery it looks like AppVeyor is not picking up the `boost-process` package I added (after I rebased on #17364

sipsorcery commented:
@Sjors the mechanism I added to save the 2 minutes updating the vcpkg source each build assumes if any installed packages are cached then all required packages are installed. As you've highlighted that's not a great assumption. 

I'll come up with a fix pronto.

Sjors commented:
Rebased after AppVeyor improvement in #17416

Sjors commented:
The Travis should go away once we ignore compiler warnings in dependencies #17520 or upgrade to Boost 1.71

Sjors commented:
I made a separate PR to bump Boost, as that seems to be causing some CI issues: #17941

Sjors commented:
Travis fails: https://travis-ci.org/bitcoin/bitcoin/jobs/641302549

```
Running 2 test cases...
2956Test cases order is shuffled using seed: 899780953
2957Entering test module "Bitcoin Core Test Suite"
2958test/system_tests.cpp(11): Entering test suite "system_tests"
2959test/system_tests.cpp(20): Entering test case "run_command"
29602020-01-24T11:25:50Z Seed: Setting random seed for current tests to RANDOM_CTX_SEED=1dbce6f9493a59dd60c8c2615905dc30901e172c156ddf6727a37f3fe178f46b
29612020-01-24T11:25:50Z Bitcoin Core version v0.19.99.0-5bc5674a (release build)
29622020-01-24T11:25:52Z Using 16 MiB out of 32/2 requested for signature cache, able to store 524288 elements
29632020-01-24T11:25:52Z Using 16 MiB out of 32/2 requested for script execution cache, able to store 524288 elements
2964Makefile:15875: recipe for target 'test/system_tests.cpp.test' failed
2965make[3]: *** [test/system_tests.cpp.test] Error 1
```

I reproduced this once on an Ubuntu machine, but after that even `git clean -dfx` doesn't reproduce.

Sjors commented:
Looks like valgrind finds an issue, but I'm not sure how to interpret the result:

```sh
valgrind --suppressions=contrib/valgrind.supp --leak-check=full src/test/test_bitcoin --run_test=system_tests
```

```
==1639== Memcheck, a memory error detector
==1639== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==1639== Using Valgrind-3.15.0 and LibVEX; rerun with -h for copyright info
==1639== Command: src/test/test_bitcoin --run_test=system_tests
==1639== 
Running 2 test cases...
==1639== 
==1639== Process terminating with default action of signal 13 (SIGPIPE)
==1639==    at 0x5492281: write (write.c:27)
==1639==    by 0x848C6E: write (basic_pipe.hpp:85)
==1639==    by 0x848C6E: _write_impl (pipe.hpp:226)
==1639==    by 0x848C6E: boost::process::basic_pipebuf<char, std::char_traits<char> >::sync() (pipe.hpp:179)
==1639==    by 0x64A4D7D: std::ostream::flush() (in /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.25)
==1639==    by 0x846A26: flush<char, std::char_traits<char> > (ostream:613)
==1639==    by 0x846A26: endl<char, std::char_traits<char> > (ostream:591)
==1639==    by 0x846A26: operator<< (ostream:113)
==1639==    by 0x846A26: runCommandParseJSON(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (system.cpp:1063)
==1639==    by 0x3C4393: system_tests::run_command::test_method() (system_tests.cpp:32)
==1639==    by 0x3C7460: system_tests::run_command_invoker() (system_tests.cpp:20)
==1639==    by 0x58F02CD: boost::detail::function::function_obj_invoker0<boost::detail::forward, int>::invoke(boost::detail::function::function_buffer&) (in /usr/lib/x86_64-linux-gnu/libboost_unit_test_framework.so.1.65.1)
==1639==    by 0x58EF77C: boost::execution_monitor::catch_signals(boost::function<int ()> const&) (in /usr/lib/x86_64-linux-gnu/libboost_unit_test_framework.so.1.65.1)
==1639==    by 0x58EF860: boost::execution_monitor::execute(boost::function<int ()> const&) (in /usr/lib/x86_64-linux-gnu/libboost_unit_test_framework.so.1.65.1)
==1639==    by 0x58EFFDC: boost::execution_monitor::vexecute(boost::function<void ()> const&) (in /usr/lib/x86_64-linux-gnu/libboost_unit_test_framework.so.1.65.1)
==1639==    by 0x591E8D0: boost::unit_test::unit_test_monitor_t::execute_and_translate(boost::function<void ()> const&, unsigned int) (in /usr/lib/x86_64-linux-gnu/libboost_unit_test_framework.so.1.65.1)
==1639==    by 0x58FAC6A: boost::unit_test::framework::state::execute_test_tree(unsigned long, unsigned int, boost::unit_test::framework::state::random_generator_helper const*) (in /usr/lib/x86_64-linux-gnu/libboost_unit_test_framework.so.1.65.1)
```

Detailed [log](https://gist.github.com/Sjors/ba14ec5c0ac27e0d77edfa40f4bbc9fd).

Sjors commented on src/util/system.cpp:
```diff
@@ -1038,6 +1042,39 @@ void runCommand(const std::string& strCommand)
+    );
+    if (!str_std_in.empty()) {
+        stdin_stream << str_std_in << std::endl;
+        stdin_stream.pipe().close();
```
Valgrind warns here: `==22385== Warning: invalid file descriptor -1 in syscall close()` during the test with `std::in`.

Sjors commented:
I added some additional checks and now Travis (with valgrind) is happy: https://travis-ci.org/bitcoin/bitcoin/builds/642905548.

Sjors committed:
[util] add RunCommandParseJSON

Sjors committed:
[depends] boost: patch unused variable in boost_process

Co-Authored-By: fanquake <fanquake@gmail.com>

Sjors commented:
Instead of bumping Boost to 1.72 I added a small patch to get rid of the unused variable in Boost Process (needed for `./configure --enable-werror`). Now that descriptor wallets are merged, this should be ready for review, as one of the prerequisites for HWI support.

fjahr reviewed (approved):
Concept ACK

Also built and ran tests with and without `--without-boost-process` on macOS.

Sjors commented on src/util/system.cpp:
```diff
@@ -1078,6 +1082,39 @@ void runCommand(const std::string& strCommand)
+
+    c.wait();
+    const int n_error = c.exit_code();
+    if (n_error) throw std::runtime_error(strprintf("runCommandParseJSON error: process(%s) returned %d\n", str_command, n_error));
```
Indeed, consumers should catch this error and do something more useful. There's different ways to do that, e.g. subclassing `std::runtime_error` or using a return enum like `TransactionError`, but we can figure that out later.

jb55 commented:
Concept ACK, weak Boost NACK but in the absence of a better alternative perhaps its fine for now. but surely it wouldn't be difficult to make a few small, well tested functions that execute external commands on each OS? It wouldn't need to be fancy or full featured like https://github.com/bitcoin/bitcoin/pull/15382#issuecomment-462370972.

Sjors commented:
> surely it wouldn't be difficult to make a few small, well tested functions that execute external commands on each OS

Maybe merging this use of Boost will motivate more people to actually and try and do that :-)

instagibbs commented:
concept ACK

MarcoFalke commented on src/test/system_tests.cpp:
```diff
@@ -0,0 +1,58 @@
+
+BOOST_FIXTURE_TEST_SUITE(system_tests, BasicTestingSetup)
+
+// At least one test is required (in case HAVE_BOOST_PROCESS is not defined)
```
Could mention that this is a workaround for https://github.com/bitcoin/bitcoin/issues/19128 ?

Sjors replied:
Done and rebased.

MarcoFalke commented:
Is it required that the signer can run arbitrary system commands in the name of bitcoind when instead some well defined interface could be defined? Or more generally, to better understand the motivation behind this pull request, I'd like to see which other options have been explored and discarded (and why).

Sjors commented:
`runCommandParseJSON` is not exposed anywhere until #16546. That PR is indeed limited to very specific commands. An earlier version of this PR did allow arbitrary commands, though only on regtest, through a `runcommand` RPC. This was dropped because of security concerns (and because I could work around it): https://github.com/bitcoin/bitcoin/pull/15382#issuecomment-549073747

Two different approaches I considered (also discussed in various places, but not sure where to find logs):
1. `bitcoind` could call an RPC instead of executing a command. That requires HWI (or similar tool) to run a server continuously, which seems problematic. It's also hard to have an RPC on localhost that can't be reached from a browser by a malicious website, so you'd have to add cookie / password authentication.
2. URI handlers. HWI could register under a URI like `bitcoin-signer:sign?transaction=PSBT` and when complete open a URI like `bitcoin-core:here_you_go?transaction=PSBT`. URI handlers might work for GUI, but they're not very suitable for RPC. They're also a pain.

If HWI (or similar tool) was malicious, it really doesn't matter if it's called through a script invocation, or if it runs as a server. It also doesn't matter what parameters you'd call it with.

MarcoFalke commented:
Fair enough. And the command is static and supplied at startup in the bitcoin.conf file (just like walletnotify), right?

Sjors commented:
@MarcoFalke yes, at least in #16546 it's passed through `-signer`. #16549 lets you configure it from the GUI without a restart, but that PR is still draft.

MarcoFalke commented:
Concept ACK

promag commented:
Tested ACK 6878673311d1081d804e2ac888473ddca3baa3a0 on macos. Are you considering supporting parsing multiline json?

fanquake commented:
Reading through the discussion here, it's still unclear to me why this is opt-out rather than opt-in. I don't understand the comment [here](https://github.com/bitcoin/bitcoin/pull/15382#issuecomment-548437890):
> On the other hand, perhaps this can be merged faster if it's off by default. In that case I'll open a separate PR to switch the default; which is necessary to make external signer functionality actually useful, rather than something only for users who self compile). But I would rather just get it over with.

We can ship releases with this on (if wanted) by adding `--enable-external-signer` to the gitian / Guix builds, and developers or users who are compiling locally can opt in when configuring. I'm not sure why having it on-by-default if Boost Process is available is a requirement for it to be "actually useful"? 

Even if it has not necessarily been the case in the past, I'd much rather have users and developers intentionally opt-in to these kinds of new features, rather than us just bundling all sorts of things into bitcoind, based on whatever you happen to have installed on the system where you are compiling.

Sjors commented:
So there's 3 scenarios:
1. binary download
2. compiled when boost::process is absent
3. compiled when boost::process is present

I mainly considered (1) and (2), because I don't think requiring self compilation for GUI features makes sense. 

If you have a strong preference for turning it off by default for (3) I can do that. However, it's unusual for features that are shipped in binaries to be turned off by default. That itself could be confusing.

Another approach could be to turn it off by default, even in the binaries, until GUI support in #16549 lands. Requiring RPC users to self-compile for now seems fine by me.

Sjors committed:
[build] make boost-process opt-in

Sjors committed:
[ci] use boost::process

Explictly opt-out on win64, in case the default changes.

jonatack commented:
Concept ACK, will review shortly

jonatack commented on src/util/system.h:
```diff
@@ -95,6 +97,16 @@ std::string ShellEscape(const std::string& arg);
+ * @param str_std_in string to pass to stdin
+ * @return parsed JSON
+ */
+UniValue runCommandParseJSON(const std::string& str_command, const std::string& str_std_in="");
```
bfed9cdd
```suggestion
UniValue RunCommandParseJSON(const std::string& str_command, const std::string& str_std_in="");
```

jonatack reviewed (commented):
Looked over the last commit bfed9cdd, the only one I felt semi-competent to review. Built with Clang on Debian with and without `--with-boost-process` and ran unit tests and just `src/test/test_bitcoin -t system_tests -l all`. I'm not sure how else to test; I did not read the middle 137 comments hidden by GitHub.

Sjors commented:
Thanks, rebased and addressed nits.

> I did not read the middle 137 comments hidden by GitHub

You don't know what you're missing :-)

Sjors commented:
A day not spent rebasing, is a day not lived... (ok, this one was trivial)

achow101 commented on src/util/system.cpp:
```diff
@@ -1079,6 +1083,39 @@ void runCommand(const std::string& strCommand)
+    bp::child c(
+        str_command,
+        bp::std_out > stdout_stream,
+        bp::std_in < stdin_stream
```
In commit `[util] add RunCommandParseJSON`

We should be piping stderr too. In HWI, this is used for some info to the user that we can discard here. But on failure, it can provide useful debugging information that we would want to see in the exception.

achow101 commented on src/test/system_tests.cpp:
```diff
@@ -0,0 +1,59 @@
+        BOOST_CHECK_EQUAL(success.getBool(), true);
+    }
+    {
+        BOOST_REQUIRE_THROW(RunCommandParseJSON("invalid_command"), std::runtime_error); // Command failed
```
In commit `[util] add RunCommandParseJSON`

Maybe use the `false` command which always fails? Or is the intent to try a command that probably won't exist?

Sjors replied:
Keeping it for now, but I suppose `false` could work too, assuming it exists on Windows.

achow101 reviewed (commented):
Some comments on configure:

* It would be nice to indicate whether we have boost process in the summary at the end of `configure`'s output.
* Instead of explicitly opting in, I would prefer boost process was used if it is available, like we do for miniupnp and zmq.

Sjors commented:
Did I mention I love rebasing? :-)

I added it to the configure summary. Also added `std::err` content to the exception message.

@achow101 wrote:

> Instead of explicitly opting in, I would prefer boost process was used if it is available, like we do for miniupnp and zmq.

So do I, but @fanquake didn't agree:

> Even if it has not necessarily been the case in the past, I'd much rather have users and developers intentionally opt-in to these kinds of new features, rather than us just bundling all sorts of things into bitcoind, based on whatever you happen to have installed on the system where you are compiling.

So for now my aim is to switch to opt-out once the GUI feature lands. In either case, I'd rather get this merged as opt-in than not at all.

promag commented:
appveyor failed with
```

C:/projects/bitcoin/src/test/system_tests.cpp(28): error: in "system_tests/run_command": check ex.what() == std::string("execve failed: No such file or directory") has failed [ CreateProcess failed: The system cannot find the file specified. != execve failed: No such file or directory]
```

> So for now my aim is to switch to opt-out once the GUI feature lands. In either case, I'd rather get this merged as opt-in than not at all.

Let's not put @Sjors's love to rebase at stake! I agree to merge this too.

MarcoFalke commented:
Looks like appveyor doesn't like this https://ci.appveyor.com/project/DrahtBot/bitcoin/builds/34093575#L811

hebasto commented:
Concept ACK.

hebasto commented on configure.ac:
```diff
@@ -1188,7 +1188,11 @@ fi
 AX_BOOST_THREAD
-AX_BOOST_PROCESS
+
+dnl Opt-in to boost-process
```
This part of the OP
> ... use `./configure --without-boost-process` to opt out...

should be inverted, right?

Sjors replied:
Fixed

hebasto commented on configure.ac:
```diff
@@ -1188,7 +1188,11 @@ fi
-AX_BOOST_PROCESS
+
+dnl Opt-in to boost-process
+if test x$with_boost_process != x; then
+  AX_BOOST_PROCESS
+fi
```
7588341b8a88370ce3382799d739063965376301
May I suggest to use [`AS_IF`](https://www.gnu.org/savannah-checkouts/gnu/autoconf/manual/autoconf-2.69/autoconf.html#index-AS_005fIF-1574) macro for [those](https://autotools.io/autoconf/m4sh.html) reasons?

hebasto reviewed (approved):
ACK fe9a6b71f47f80d0a77d239948341ea168d600d0, tested on Linux Mint 20 (x86_64).

nits:
- pair `#ifdef HAVE_BOOST_PROCESS` with `#endif // HAVE_BOOST_PROCESS`
- https://github.com/bitcoin/bitcoin/pull/15382#discussion_r456932313
- https://github.com/bitcoin/bitcoin/pull/15382#discussion_r456930644

Sjors commented:
Thanks @hebasto, I'll apply those suggestions if I need to rebase again.

Sjors commented:
Rebased and addressed @hebasto's comments.

hebasto commented on depends/packages/boost.mk:
```diff
@@ -31,7 +31,9 @@ $(package)_cxxflags_linux=-fPIC
 
+# Fix unused variable in boost_process, can be removed after upgrading to 1.72
 define $(package)_preprocess_cmds
+	sed -i.old "s/int ret_sig = 0;//" boost/process/detail/posix/wait_group.hpp && \
```
Is a tab character placed here intentionally?

Sjors replied:
Just copy-pasted that line... Fixed, since I had to rebase again.

hebasto reviewed (approved):
re-ACK 7d88848793b10aafd530e6eb95fbdb19570f6819, only rebased (verified with `git range-diff`) and suggested changes implemented since the [previous](https://github.com/bitcoin/bitcoin/pull/15382#pullrequestreview-451148846) review.

hebasto reviewed (approved):
re-ACK 31cf68a3ad1f0a5537c8419e2912b55fbfb88fa0, only rebased (verified with `git range-diff`) and removed an unintentional tab character since the [previous](https://github.com/bitcoin/bitcoin/pull/15382#pullrequestreview-458371035) review.

Sjors commented:
I think the sanatizer failure is unrelated. Restarted. Log for reference: https://cirrus-ci.com/task/5039271566376960

achow101 commented:
ACK 31cf68a3ad1f0a5537c8419e2912b55fbfb88fa0

meshcollider reviewed (commented):
Very light utACK 31cf68a3ad1f0a5537c8419e2912b55fbfb88fa0, although I am not very confident with build stuff.

promag reviewed (commented):
Code review ACK 31cf68a3ad1f0a5537c8419e2912b55fbfb88fa0, don't mind the nit.

ryanofsky commented on src/test/system_tests.cpp:
```diff
@@ -0,0 +1,95 @@
+        const UniValue result = RunCommandParseJSON("echo '{\"success\": true}'");
+#else
+        // ... but Linux and macOS echo a single quote if it's used
+        const UniValue result = RunCommandParseJSON("echo \"{\"success\": true}\"");
```
In commit "[util] add RunCommandParseJSON" (31cf68a3ad1f0a5537c8419e2912b55fbfb88fa0)

This code seems nonsensical, and I spent like an hour trying to figure out how it might be working. I think what is happening is that boost's [build_arg](https://github.com/boostorg/process/blob/4fd88876013a9547b8bc7183d0096c86e9a1141f/include/boost/process/detail/posix/basic_cmd.hpp#L53) function is building this up into a vector of three strings:

- `echo`
- `"{"success":`
- `true}"`

and passing these as `argv` to [execve](https://linux.die.net/man/2/execve), which I guess gets the job done but is not the most straightforward thing. Might be better to pass an argument vector, or a normal bash command line instead of a fake command line string that uses boost's weird command line splitting code and escaping rules. This would just be for clarity though, since I think the current code probably should work.

On `#ifdef WIN32` code, I don't see how that code is working, and I wonder it is actually running and passing on windows with appveyor, or is just broken and should be removed. Since we don't seem to be enabling the shell (cmd.exe) as a boost process option, and since single quotes have no meaning on windows, and since windows doesn't have an echo.exe binary, it seems impossible that this could be working.

Sjors replied:
"boost's weird command line splitting code and escaping rule" are what the user has to deal with when using `-signer=` in the followup PR, so it seems reasonable to have to test use that too.

I fixed a failed AppVeyor tests with this `#ifdef WIN32` line, so the code does run. The external signer Python tests also run on AppVeyor in the followup PR.  Maybe it's running in the power shell?

Sjors commented on build-aux/m4/ax_boost_process.m4:
```diff
@@ -0,0 +1,130 @@
+                [[boost::process::child* child = new boost::process::child; delete child;]])],
+                ax_cv_boost_process=yes, ax_cv_boost_process=no)
+			 CXXFLAGS=$CXXFLAGS_SAVE
+             AC_LANG_POP([C++])
```
Any lasting changes to this file would have to made upstream: https://www.gnu.org/software/autoconf-archive/ax_boost_process.html

ryanofsky reviewed (approved):
Code review ACK 31cf68a3ad1f0a5537c8419e2912b55fbfb88fa0. I left some comments below that could be ignored or followed up later. The current change is clean and comprehensive.
